File: hexapod/hexapod_ros/hexapod_ros/src/leg.py
from .parameter import *
import traceback

class Leg:
    """
    Handles Inverse Kinematics (IK) for a single leg.
    Translates (X, Y, Z) coordinates into (Coxa, Femur, Tibia) servo angles.
    """
    def __init__(self, leg):
        self.leg = leg # 0 ~ 5, No. of leg, reference to graph in parameter file
        # Foot position relative to leg base
        self.x, self.y, self.z = x_offset, y_offset, z_offset
        # Joint angles (coxa, femur, tibia)
        self.a, self.b, self.c = 0, 0, 180
        self.offset_angle = offset_angle_map[self.leg]
        self.IK()

    def calculateWalk(self, phase, direction, distance, pitch, roll, yaw, rotate):
        """ Swing motion, leg tip will draw a semicircle, motion integrated with balance"""
        if phase <= 180:
            turn_distance = 0.5 * distance
            x = direction[0] * (turn_distance * cos(radians(180 - phase)) + turn_distance)
            y = direction[1] * - (turn_distance * cos(radians(180 - phase)) + turn_distance) 
            self.z = - distance * sin(radians(phase)) + z_offset
        else:
            turn_distance = distance
            x = direction[0] * (-1 * turn_distance * ((phase/180) - 1) + distance)
            y = direction[1] * - (-1 * turn_distance * ((phase/180) - 1) + distance)
            self.z = z_offset
        if not rotate: new_vec = transformLegtoBody(self.leg, [x,y])
        else: new_vec = [x, y]
        self.x = x_offset + new_vec[0]
        self.y = y_offset + new_vec[1]
        R_c = R(self.offset, self.offset_angle, roll, pitch, yaw, [self.x, self.y]) 
        if phase > 180:
            roll_vec = transformLegtoBody(self.leg, R_c[0][:2])
            pitch_vec = transformLegtoBody(self.leg, R_c[1][:2])
            self.x += roll_vec[0] + pitch_vec[0]
            self.y += roll_vec[1] + pitch_vec[1]
            self.z += R_c[0][2] + R_c[1][2]
        yaw_vec = transformLegtoBody(self.leg, R_c[2][:2])
        self.x += yaw_vec[0]
        self.y += yaw_vec[1]
        self.IK()
        self.verify_IK()
        self.angleToDC()
        return (self.a, self.b, self.c)

    def stand(self,l,c): # stand up function (brute force)
        self.IK()
        d = 20 - (20/l)*c
        self.b += d
        self.c += d
        self.verify_IK()
        self.angleToDC()
        return (self.a, self.b, self.c)

    def balance(self, pitch, roll): # only roll & pitch, no yaw
        R_c = R(self.offset, self.offset_angle, roll, pitch) 
        roll_vec = transformLegtoBody(self.leg, R_c[0][:2])
        pitch_vec = transformLegtoBody(self.leg, R_c[1][:2])
        self.x = x_offset + roll_vec[0] + pitch_vec[0]
        self.y = y_offset + roll_vec[1] + pitch_vec[1]
        self.z = z_offset + R_c[0][2] + R_c[1][2]
        self.IK()
        self.verify_IK()
        self.angleToDC()
        return (self.a, self.b, self.c)
    
    @staticmethod
    def FK(joint_angles, DH_table):
        # find end effector positoin given joint angles
        theta1, theta2, theta3 = joint_angles
        H0_1 = H(DH_table[0][0], DH_table[0][1], DH_table[0][2], radians(theta1))
        H1_2 = H(DH_table[1][0], DH_table[1][1], DH_table[1][2], radians(theta2))
        H2_3 = H(DH_table[2][0], DH_table[2][1], DH_table[2][2], radians(theta3))
        H0_2 = H0_1@H1_2
        H0_3 = H0_2@H2_3
        x, y, z = H0_3[0][3], H0_3[1][3], H0_3[2][3]  # extract position vector
        return (x, y, z)   
    
    def verify_IK(self):
        target = (self.x, self.y, self.z)
        fk_pos = Leg.FK((self.a, self.b, self.c), DH_table)
        error = (
            fk_pos[0] - target[0],
            fk_pos[1] - target[1],
            fk_pos[2] - target[2]
        )

        print("Target:", target)
        print("FK:", fk_pos)
        print("Error:", error)

        if any(abs(e) > 1e-2 for e in error):
            print("IK verification failed")
   
    def IK(self):
        """
        Inverse Kinematics: 
        Calculates servo angles (self.a, self.b, self.c) from desired (self.x, self.y, self.z) foot position.
        """
        try:
            y = sqrt(self.x**2 + self.y**2) - cl # actual location of tip from femur in coxa frame
            k = sqrt(y**2 + self.z**2) # shortest distance from femur to tibia tip 
            
            # Constrain k to reachable workspace
            k = max(abs(fl - tl), min(fl + tl, k))
            
            theta = asin(max(-1, min(1, self.z / k))) # angle between horizontal and k
            theta2 = acos(max(-1, min(1, ((fl**2)+(k**2)-(tl**2))/(2*fl*k)))) # cosine angle of tibia

            self.a = max(min(coxa_range[1], degrees(atan2(self.x, self.y))), coxa_range[0]) # coxa angle
            self.b = max(min(femur_range[1], degrees(theta2 - theta)), femur_range[0]) # femur angle
            self.c = max(min(tibia_range[1], 180 - degrees(acos(max(-1, min(1, ((fl**2)+(tl**2)-(k**2))/(2*fl*tl)))))), tibia_range[0]) # tibia angle from femur
        except Exception:
            traceback.print_exc()
            self.a = self.b = self.c = 0
        
    def angleToDC(self):
        """Converts raw degrees to Servo Duty Cycle (PWM) values. (4000 ~ 8000)"""
        self.a = int((100 * self.a)/3 + 6000)
        self.b = int((100 * self.b)/3 + 6000)
        self.c = int((100 * self.c)/3 + 3000)

    def curr_angle(self):
        return (self.a, self.b, self.c)

File: hexapod/firmware/src/kinematics/leg.cpp
#include "leg.h"
#include <math.h>
#include "parameter.h"

// parameters
float cl = 55.65;
float fl = 110.0;
float tl = 145.4;

// Neutral position when leg is horizontally straight is (0, 312, 0)
float x_offset = 0;
float y_offset = 312;
float z_offset = 0;

float coxa_min = 135;
float coxa_max = 225;

float femur_min = 70;
float femur_max = 330;

float tibia_min = 60;
float tibia_max = 300;

float clampf(float val, float minv, float maxv) {
  if (val < minv) return minv;
  if (val > maxv) return maxv;
  return val;
}

Leg::Leg() {
  x = x_offset;
  y = y_offset;
  z = z_offset;
}

void Leg::setFoot(float nx, float ny, float nz) {
  x = nx;
  y = ny;
  z = nz;
  IK();
}

void Leg::IK() {

  float y1 = sqrt(x*x + y*y) - cl;
  float k  = sqrt(y1*y1 + z*z);

  k = clampf(k, fabs(fl - tl), fl + tl);

  float ratio1 = clampf(z / k, -1.0, 1.0);
  float theta = asin(ratio1);

  float ratio2 = clampf((fl*fl + k*k - tl*tl) / (2*fl*k), -1.0, 1.0);
  float theta2 = acos(ratio2);

  float ratio3 = clampf((fl*fl + tl*tl - k*k) / (2*fl*tl), -1.0, 1.0);
  float tib = acos(ratio3);

  // IK angles
  float ia = atan2(x, y) * RAD_TO_DEG;
  float ib = (theta2 - theta) * RAD_TO_DEG;
  float ic = 180 - tib * RAD_TO_DEG;

  // map to servo reference frame, +180 because we set 180 to be neutral for ST3215
  a = 180.0 + ia;
  b = 180.0 + ib;
  c = 180.0 + ic;

  // safety clamps
  a = clampf(a, coxa_min, coxa_max);
  b = clampf(b, femur_min, femur_max);
  c = clampf(c, tibia_min, tibia_max);
}

void Leg::FK(float a_in, float b_in, float c_in, float &fx, float &fy, float &fz)
{
  // Convert servo angles back to IK reference
  float ia = a_in - 180;
  float ib = b_in - 180;
  float ic = c_in - 180;

  float tib = 180 - ic;

  float theta = ib * DEG_TO_RAD;
  float tib_rad = tib * DEG_TO_RAD;

  float k = sqrt(fl*fl + tl*tl - 2*fl*tl*cos(tib_rad));

  float theta2 = acos((fl*fl + k*k - tl*tl)/(2*fl*k));

  float z = k * sin(theta2 - theta);
  float y1 = k * cos(theta2 - theta);

  float planar = y1 + cl;

  fx = planar * sin(ia * DEG_TO_RAD);
  fy = planar * cos(ia * DEG_TO_RAD);
  fz = z;
}

void Leg::verifyIK()
{
  float fx, fy, fz;

  FK(a, b, c, fx, fy, fz);

  Serial.print("Target: ");
  Serial.print(x); Serial.print(" ");
  Serial.print(y); Serial.print(" ");
  Serial.println(z);

  Serial.print("FK: ");
  Serial.print(fx); Serial.print(" ");
  Serial.print(fy); Serial.print(" ");
  Serial.println(fz);

  Serial.print("Error: ");
  Serial.print(fx - x); Serial.print(" ");
  Serial.print(fy - y); Serial.print(" ");
  Serial.println(fz - z);
}

void Leg::stand(int steps, int current_step)
{
  IK();

  float d = 20 - (20.0 / steps) * current_step;

  b += d;
  c += d;

  b = clampf(b, femur_min, femur_max);
  c = clampf(c, tibia_min, tibia_max);
}

void Leg::balance(float pitch, float roll)
{
  float dx = pitch * 2;
  float dy = roll * 2;

  x = x_offset + dx;
  y = y_offset + dy;
  z = z_offset;

  IK();
}

void Leg::balance(float pitch, float roll)
{
  float dx = pitch * 2;
  float dy = roll * 2;

  x = x_offset + dx;
  y = y_offset + dy;
  z = z_offset;

  IK();
}

void Leg::printAngles() {
  Serial.print("  Coxa: ");
  Serial.print(a);
  Serial.print("  Femur: ");
  Serial.print(b);
  Serial.print("  Tibia: ");
  Serial.println(c);
}
